src/libfreeswan/initaddr.3
.TH IPSEC_INITADDR 3 "11 Sept 2000"
.SH NAME
ipsec initaddr \- initialize an ip_address
.br
ipsec addrtypeof \- get address type of an ip_address
.br
ipsec addrlenof \- get length of address within an ip_address
.br
ipsec addrbytesof \- get copy of address within an ip_address
.br
ipsec addrbytesptr \- get pointer to address within an ip_address
.SH SYNOPSIS
.B "#include <freeswan.h>"
.sp
.B "const char *initaddr(const char *src, size_t srclen,"
.ti +1c
.B "int af, ip_address *dst);"
.br
.B "int addrtypeof(const ip_address *src);"
.br
.B "size_t addrlenof(const ip_address *src);"
.br
.B "size_t addrbytesof(const ip_address *src,"
.ti +1c
.B "unsigned char *dst, size_t dstlen);"
.br
.B "size_t addrbytesptr(const ip_address *src,"
.ti +1c
.B "const unsigned char **dst);"
.SH DESCRIPTION
The
.B <freeswan.h>
library uses an internal type
.I ip_address
to contain one of the (currently two) types of IP address.
These functions provide basic tools for creating and examining this type.
.PP
.I Initaddr
initializes a variable
.I *dst
of type
.I ip_address
from an address
(in network byte order,
indicated by a pointer
.I src
and a length
.IR srclen )
and an address family
.I af
(typically
.B AF_INET
or
.BR AF_INET6 ).
The length must be consistent with the address family.
.PP
.I Addrtypeof
returns the address type of an address,
normally
.B AF_INET
or
.BR AF_INET6 .
(The
.B <freeswan.h>
header file arranges to include the necessary headers for these
names to be known.)
.PP
.I Addrlenof
returns the size (in bytes) of the address within an
.IR ip_address ,
to permit storage allocation etc.
.PP
.I Addrbytesof
copies the address within the
.I ip_address
.I src
to the buffer indicated by the pointer
.I dst
and the length
.IR dstlen ,
and returns the address length (in bytes).
If the address will not fit,
as many bytes as will fit are copied;
the returned length is still the full length.
It is the caller's responsibility to check the
returned value to ensure that there was enough room.
.PP
.I Addrbytesptr
sets
.I *dst
to a pointer to the internal address within the
.IR ip_address ,
and returns the address length (in bytes).
If
.I dst
is
.BR NULL ,
it just returns the address length.
The pointer points to
.B const
to discourage misuse.
.PP
.I Initaddr
returns
.B NULL
for success and
a pointer to a string-literal error message for failure;
see DIAGNOSTICS.
.PP
The functions which return
.I size_t
return
.B 0
for a failure.
.SH SEE ALSO
inet(3), ipsec_ttoaddr(3)
.SH DIAGNOSTICS
An unknown address family is a fatal error for any of these functions
except
.IR addrtypeof .
An address-size mismatch is a fatal error for
.IR initaddr .
.SH HISTORY
Written for the FreeS/WAN project by Henry Spencer.
.SH BUGS
.I Addrtypeof
should probably have been named
.IR addrfamilyof .


NEWS
strongswan-4.3.1
----------------

- The nm plugin now passes DNS/NBNS server information to NetworkManager,
  allowing Gateway administrator to set DNS/NBNS configuration on clients
  dynamically.

strongswan-4.3.0
----------------

- Support for the IKEv2 Multiple Authentication Exchanges extension (RFC4739).
  Initiators and responders can use several authentication rounds (e.g. RSA
  followed by EAP) to authenticate. The new ipsec.conf leftauth/rightauth and
  leftauth2/rightauth2 parameters define own authentication rounds or setup
  constraints for the remote peer. See the ipsec.conf man page for more detials.

- If glibc printf hooks (register_printf_function) are not available,
  strongSwan can use the vstr string library to run on non-glibc systems.

- The IKEv2 charon daemon can now configure the ESP CAMELLIA-CBC cipher
  (esp=camellia128|192|256).

- Refactored the pluto and scepclient code to use basic functions (memory
  allocation, leak detective, chunk handling, printf_hooks, strongswan.conf
  attributes, ASN.1 parser, etc.) from the libstrongswan library.

- Up to two DNS and WINS servers to be sent via IKEv1 ModeConfig can be
  configured in the pluto section of strongswan.conf.


strongswan-4.2.14
-----------------

- The new server-side EAP RADIUS plugin (--enable-eap-radius)
  relays EAP messages to and from a RADIUS server. Succesfully
  tested with with a freeradius server using EAP-MD5 and EAP-SIM.

- A vulnerability in the Dead Peer Detection (RFC 3706) code was found by
  Gerd v. Egidy <gerd.von.egidy@intra2net.com> of Intra2net AG affecting
  all Openswan and strongSwan releases. A malicious (or expired ISAKMP)
  R_U_THERE or R_U_THERE_ACK Dead Peer Detection packet can cause the
  pluto IKE daemon to crash and restart. No authentication or encryption
  is required to trigger this bug. One spoofed UDP packet can cause the
  pluto IKE daemon to restart and be unresponsive for a few seconds while
  restarting. This DPD null state vulnerability has been officially
  registered as CVE-2009-0790 and is fixed by this release.

- ASN.1 to time_t conversion caused a time wrap-around for
  dates after Jan 18 03:14:07 UTC 2038 on 32-bit platforms.
  As a workaround such dates are set to the maximum representable
  time, i.e. Jan 19 03:14:07 UTC 2038.

- Distinguished Names containing wildcards (*) are not sent in the
  IDr payload anymore. 


strongswan-4.2.13
-----------------

- Fixed a use-after-free bug in the DPD timeout section of the
  IKEv1 pluto daemon which sporadically caused a segfault.

- Fixed a crash in the IKEv2 charon daemon occuring with
  mixed RAM-based and SQL-based virtual IP address pools. 

- Fixed ASN.1 parsing of algorithmIdentifier objects where the
  parameters field is optional.

- Ported nm plugin to NetworkManager 7.1.


strongswan-4.2.12
-----------------

- Support of the EAP-MSCHAPv2 protocol enabled by the option
  --enable-eap-mschapv2. Requires the MD4 hash algorithm enabled
  either by --enable-md4 or --enable-openssl.

- Assignment of up to two DNS and up to two WINS servers to peers via
  the IKEv2 Configuration Payload (CP). The IPv4 or IPv6 nameserver 
  addresses are defined in strongswan.conf.

- The strongSwan applet for the Gnome NetworkManager is now built and
  distributed as a separate tarball under the name NetworkManager-strongswan.

 
strongswan-4.2.11
-----------------

- Fixed ESP NULL encryption broken by the refactoring of keymat.c.
  Also introduced proper initialization and disposal of keying material.

- Fixed the missing listing of connection definitions in ipsec statusall
  broken by an unfortunate local variable overload.


strongswan-4.2.10
-----------------

- Several performance improvements to handle thousands of tunnels with almost
  linear upscaling. All relevant data structures have been replaced by faster
  counterparts with better lookup times.

- Better parallelization to run charon on multiple cores. Due to improved
  ressource locking and other optimizations the daemon can take full
  advantage of 16 or even more cores.

- The load-tester plugin can use a NULL Diffie-Hellman group and simulate
  unique identities and certificates by signing peer certificates using a CA
  on the fly.

- The redesigned stroke in-memory IP pool handles leases. The "ipsec leases"
  command queries assigned leases.

- Added support for smartcards in charon by using the ENGINE API provided by
  OpenSSL, based on patches by Michael Roßberg.

- The Padlock plugin supports the hardware RNG found on VIA CPUs to provide a
  reliable source of randomness.

strongswan-4.2.9
----------------

- Flexible configuration of logging subsystem allowing to log to multiple
  syslog facilities or to files using fine-grained log levels for each target.

- Load testing plugin to do stress testing of the IKEv2 daemon against self
  or another host. Found and fixed issues during tests in the multi-threaded
  use of the OpenSSL plugin.

- Added profiling code to synchronization primitives to find bottlenecks if
  running on multiple cores. Found and fixed an issue where parts of the
  Diffie-Hellman calculation acquired an exclusive lock. This greatly improves
  parallelization to multiple cores.

- updown script invocation has been separated into a plugin of its own to
  further slim down the daemon core.

- Separated IKE_SA/CHILD_SA key derivation process into a closed system,
  allowing future implementations to use a secured environment in e.g. kernel
  memory or hardware.

- The kernel interface of charon has been modularized. XFRM NETLINK (default)
  and PFKEY (--enable-kernel-pfkey) interface plugins for the native IPsec
  stack of the Linux 2.6 kernel as well as a PFKEY interface for the KLIPS
  IPsec stack (--enable-kernel-klips) are provided.

- Basic Mobile IPv6 support has been introduced, securing Binding Update
  messages as well as tunneled traffic between Mobile Node and Home Agent.

src/libfreeswan/initaddr.c
/*
 * initialize address structure
 * Copyright (C) 2000  Henry Spencer.
 * 
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.  See <http://www.fsf.org/copyleft/lgpl.txt>.
 * 
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
 * License for more details.
 */
#include <sys/socket.h>

#include "internal.h"
#include "freeswan.h"

/*
 - initaddr - initialize ip_address from bytes
 */
err_t				/* NULL for success, else string literal */
initaddr(src, srclen, af, dst)
const unsigned char *src;
size_t srclen;
int af;				/* address family */
ip_address *dst;
{
	switch (af) {
	case AF_INET:
		if (srclen != 4)
			return "IPv4 address must be exactly 4 bytes";
		dst->u.v4.sin_family = af;
		dst->u.v4.sin_port = 0;		/* unused */
		memcpy((char *)&dst->u.v4.sin_addr.s_addr, src, srclen);
		break;
	case AF_INET6:
		if (srclen != 16)
			return "IPv6 address must be exactly 16 bytes";
		dst->u.v6.sin6_family = af;
		dst->u.v6.sin6_flowinfo = 0;		/* unused */
		dst->u.v6.sin6_port = 0;		/* unused */
		memcpy((char *)&dst->u.v6.sin6_addr, src, srclen);
		break;
	default:
		return "unknown address family in initaddr";
		break;
	}
	return NULL;
}
